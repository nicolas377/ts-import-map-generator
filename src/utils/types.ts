export const enum OptionType {
  Help,
  Version,
}
