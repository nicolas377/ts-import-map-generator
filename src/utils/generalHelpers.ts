export type AnyFunction = (...args: never[]) => void;

export const newLineCharacter = "\n";
export const emptyArray = [];

// eslint-disable-next-line @typescript-eslint/no-empty-function
export function noopFunction(): void {}

export const enum Comparison {
  LessThan = -1,
  Equal = 0,
  GreaterThan = 1,
}

export function compareValues(left: number, right: number): Comparison {
  if (left < right) return Comparison.LessThan;
  if (left > right) return Comparison.GreaterThan;
  return Comparison.Equal;
}

// TODO: use plain Array.isArray once ts updates the Array.isArray defs to support narrowing readonly arrays
export function checkIsReadonlyArray(
  value: unknown
): value is readonly unknown[] {
  return Array.isArray(value);
}

export function keysOfObject<T extends PropertyKey>(
  obj: Record<T, unknown>
): T[] {
  return Object.keys(obj) as T[];
}

export function entriesOfObject<T extends PropertyKey, U>(
  obj: Record<T, U>
): [T, U][] {
  return Object.entries(obj) as [T, U][];
}

export function objectHasProperty<T extends PropertyKey>(
  record: Readonly<Record<T, unknown>>,
  key: PropertyKey
): key is T {
  return Object.prototype.hasOwnProperty.call(record, key);
}

export function arrayIncludesValue<T>(
  arr: readonly T[],
  value: unknown
): value is T {
  return arr.includes(value as T);
}

// production build detection

export const productionBuild =
  process.env.IMG_PROJECT_ENV === "production" ||
  // if the environment variable is set, don't check for ts-node
  (typeof process.env.IMG_PROJECT_ENV === "undefined" &&
    typeof (process as unknown as Record<PropertyKey, unknown>)[
      Symbol.for("ts-node.register.instance")
    ] !== "undefined");

export const isCli = process.env.IMG_PROJECT_RUN_TYPE === "cli";
