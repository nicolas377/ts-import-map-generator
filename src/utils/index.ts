export * from "utils/debug";
export * from "utils/generalHelpers";
export * from "utils/options";
