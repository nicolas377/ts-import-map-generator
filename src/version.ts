import { version as programVersion } from "../package.json";
import {
  checkIsReadonlyArray,
  compareValues,
  Comparison,
  Debug,
  emptyArray,
} from "utils";

// Implements the SemVer 2.0 spec
// A lot of this code is borrowed from typescript's semver.ts, but repurposed for our purposes
export class Semver {
  static readonly zero = new Semver(0, 0, 0, ["0"]);

  public readonly major: number;
  public readonly minor: number;
  public readonly patch: number;
  public readonly prerelease: readonly string[];
  public readonly build: readonly string[];

  constructor(text: string);
  constructor(
    major: number,
    minor: number,
    patch: number,
    prerelease?: string | readonly string[],
    build?: string | readonly string[]
  );
  constructor(
    major: number | string,
    // These defaults are never used, they're just here to provide type compatibility between the signatures
    minor = 0,
    patch = 0,
    prerelease: string | readonly string[] = "",
    build: string | readonly string[] = ""
  ) {
    if (typeof major === "string") {
      const result = Debug.checkDefined(
        parseSemverComponents(major),
        "Invalid version"
      );

      ({ major, minor, patch, prerelease, build } = result);
    }

    Debug.assert(
      !Number.isNaN(major) && major >= 0,
      "Major version should be a positive number"
    );
    Debug.assert(
      !Number.isNaN(minor) && minor >= 0,
      "Minor version should be a positive number"
    );
    Debug.assert(
      !Number.isNaN(patch) && patch >= 0,
      "Patch version should be a positive number"
    );

    const prereleaseArray =
      prerelease === ""
        ? emptyArray
        : checkIsReadonlyArray(prerelease)
        ? prerelease
        : prerelease.split(".");
    const buildArray =
      build === ""
        ? emptyArray
        : checkIsReadonlyArray(build)
        ? build
        : build.split(".");

    this.major = major;
    this.minor = minor;
    this.patch = patch;
    this.prerelease = prereleaseArray;
    this.build = buildArray;
  }

  public toString(): string {
    let result = `${this.major}.${this.minor}.${this.patch}`;
    if (this.prerelease.length) result += `-${this.prerelease.join(".")}`;
    if (this.build.length) result += `+${this.build.join(".")}`;
    return result;
  }

  private comparePrerelease(
    left: readonly string[],
    right: readonly string[]
  ): Comparison {
    if (left === right) return Comparison.Equal;
    if (left.length === 0)
      return right.length === 0 ? Comparison.Equal : Comparison.GreaterThan;
    if (right.length === 0) return Comparison.LessThan;

    // The earlier elements are more significant, and comparisons from them are returned if they're not equal.
    const length = Math.min(left.length, right.length);
    for (let i = 0; i < length; i++) {
      const leftIdentifier = left[i]!;
      const rightIdentifier = right[i]!;

      if (leftIdentifier === rightIdentifier) continue;

      const leftIsNumber = stringIsNumeric(leftIdentifier);
      const rightIsNumber = stringIsNumeric(rightIdentifier);

      if (leftIsNumber || rightIsNumber) {
        // Non-numeric values get precedence over numeric values
        if (leftIsNumber !== rightIsNumber) {
          return leftIsNumber ? Comparison.LessThan : Comparison.GreaterThan;
        }

        // Identifiers consisting of only digits are compared numerically
        const result = compareValues(+leftIdentifier, +rightIdentifier);
        if (result !== Comparison.Equal) return result;
      }
    }

    // Larger sets of prerelease components get precedence.
    return compareValues(left.length, right.length);

    function stringIsNumeric(s: string) {
      return !Number.isNaN(+s);
    }
  }

  // remember that the result says that this instance is blank to the argument
  public compareTo(other: Semver | undefined): Comparison {
    if (other === undefined) {
      return Comparison.GreaterThan;
    }
    if (this === other) {
      return Comparison.Equal;
    }

    return (
      compareValues(this.major, other.major) ||
      compareValues(this.minor, other.minor) ||
      compareValues(this.patch, other.patch) ||
      this.comparePrerelease(this.prerelease, other.prerelease)
    );
  }
}

interface SemverComponents {
  major: number;
  minor: number;
  patch: number;
  prerelease: string;
  build: string;
}

function parseSemverComponents(text: string): SemverComponents | undefined {
  const match =
    /^(0|[1-9]\d*)(?:\.(0|[1-9]\d*)(?:\.(0|[1-9]\d*)(?:-([a-z0-9-.]+))?(?:\+([a-z0-9-.]+))?)?)?$/i.exec(
      text
    );

  if (!match) return undefined;

  const [, major, minor = "0", patch = "0", prerelease = "", build = ""] =
    match;

  Debug.assertIsDefined(major, "Major version should be defined");

  return {
    major: parseInt(major, 10),
    minor: parseInt(minor, 10),
    patch: parseInt(patch, 10),
    prerelease,
    build,
  };
}

export const generatorVersion = new Semver(programVersion);

export { programVersion };
