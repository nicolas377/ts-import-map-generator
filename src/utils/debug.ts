import {
  AnyFunction,
  Comparison,
  noopFunction,
  objectHasProperty,
} from "./generalHelpers";
import { generatorVersion, Semver } from "version";

export interface LoggingHost {
  log(level: LogLevel, s: string): void;
}

export const enum LogLevel {
  Off,
  Error,
  Warning,
  Info,
  Verbose,
}

export namespace Debug {
  // logging

  // eslint-disable-next-line prefer-const
  export let currentLogLevel = LogLevel.Warning;
  export let loggingHost: LoggingHost | undefined;

  function shouldLog(level: LogLevel): boolean {
    return currentLogLevel <= level;
  }

  function logMessage(level: LogLevel, s: string) {
    if (loggingHost && shouldLog(level)) {
      loggingHost.log(level, s);
    }
  }

  export namespace log {
    export function error(s: string) {
      logMessage(LogLevel.Error, s);
    }

    export function warn(s: string): void {
      logMessage(LogLevel.Warning, s);
    }

    export function info(s: string): void {
      logMessage(LogLevel.Info, s);
    }

    export function trace(s: string): void {
      logMessage(LogLevel.Verbose, s);
    }
  }

  // assertions

  export function fail(message?: string, stackCrawlMark?: AnyFunction): never {
    // This is a great place to put a breakpoint; whenever we fail, we'll stop here.
    const e = new Error(
      message ? `Debug Failure. ${message}` : "Debug Failure."
    );

    Error.captureStackTrace(e, stackCrawlMark || fail);

    throw e;
  }

  export function assert(
    expression: unknown,
    message?: string,
    stackCrawlMark?: AnyFunction
  ): asserts expression {
    if (!expression) {
      fail(
        message ? `False expression: ${message}` : "False expression.",
        stackCrawlMark || assert
      );
    }
  }

  export function assertEqual<T>(
    a: T,
    b: T,
    msg?: string,
    msg2?: string,
    stackCrawlMark?: AnyFunction
  ) {
    if (a !== b) {
      const message = msg ? (msg2 ? `${msg} ${msg2}` : msg) : "";
      fail(`Expected ${a} === ${b}. ${message}`, stackCrawlMark || assertEqual);
    }
  }

  export function assertLessThan(
    a: number,
    b: number,
    msg?: string,
    stackCrawlMark?: AnyFunction
  ): void {
    if (a >= b) {
      fail(
        `Expected ${a} < ${b}. ${msg || ""}`,
        stackCrawlMark || assertLessThan
      );
    }
  }

  export function assertLessThanOrEqual(
    a: number,
    b: number,
    stackCrawlMark?: AnyFunction
  ): void {
    if (a > b) {
      fail(`Expected ${a} <= ${b}`, stackCrawlMark || assertLessThanOrEqual);
    }
  }

  export function assertGreaterThanOrEqual(
    a: number,
    b: number,
    stackCrawlMark?: AnyFunction
  ): void {
    if (a < b) {
      fail(`Expected ${a} >= ${b}`, stackCrawlMark || assertGreaterThanOrEqual);
    }
  }

  export function assertIsDefined<T>(
    value: T,
    message?: string,
    stackCrawlMark?: AnyFunction
  ): asserts value is NonNullable<T> {
    if (value === undefined || value === null) {
      fail(message, stackCrawlMark || assertIsDefined);
    }
  }

  export function checkDefined<T>(
    value: T | null | undefined,
    message?: string,
    stackCrawlMark?: AnyFunction
  ): T {
    assertIsDefined(value, message, stackCrawlMark || checkDefined);
    return value;
  }

  export function assertEachIsDefined<T>(
    value: readonly T[],
    message?: string,
    stackCrawlMark?: AnyFunction
  ): asserts value is readonly NonNullable<T>[] {
    for (const v of value) {
      assertIsDefined(v, message, stackCrawlMark || assertEachIsDefined);
    }
  }

  export function assertNever(
    value: never,
    message = "Expected function to never be called",
    stackCrawlMark?: AnyFunction
  ): never {
    return fail(message, stackCrawlMark || assertNever);
  }

  // deprecations

  interface DeprecationOptions {
    message?: string;
    error?: boolean;
    since?: Semver | string;
    warnAfter?: Semver | string;
    errorAfter?: Semver | string;
    name?: string;
  }

  function getFunctionName(func: AnyFunction) {
    if (typeof func !== "function") {
      return "";
    } else if (objectHasProperty(func, "name")) {
      return func.name;
    } else {
      const text = Function.prototype.toString.call(func);
      const match = /^function\s+([\w$]+)\s*\(/.exec(text);
      return match ? match[1]! : "";
    }
  }

  function formatDeprecationMessage(
    name: string,
    error: boolean,
    errorAfter: Semver | undefined,
    since: Semver | undefined,
    message: string | undefined
  ): string {
    const deprecationMessage =
      (error ? "DeprecationError: " : "DeprecationWarning: ") +
      `'${name}' ` +
      (since ? `has been deprecated since v${since}` : "is deprecated") +
      (error
        ? " and can no longer be used."
        : errorAfter
        ? ` and will no longer be usable after v${errorAfter}`
        : ".") +
      (message ?? "");

    return deprecationMessage;
  }

  function createErrorDeprecation(
    name: string,
    errorAfter: Semver | undefined,
    since: Semver | undefined,
    message: string | undefined
  ) {
    const deprecationMessage = formatDeprecationMessage(
      name,
      true,
      errorAfter,
      since,
      message
    );

    return () => {
      throw new TypeError(deprecationMessage);
    };
  }

  function createWarningDeprecation(
    name: string,
    errorAfter: Semver | undefined,
    since: Semver | undefined,
    message: string | undefined
  ) {
    let hasWrittenDeprecation = false;

    return () => {
      if (!hasWrittenDeprecation) {
        log.warn(
          formatDeprecationMessage(name, false, errorAfter, since, message)
        );
        hasWrittenDeprecation = true;
      }
    };
  }

  function createDeprecation(
    name: string,
    options: DeprecationOptions & { error: true }
  ): () => never;
  function createDeprecation(
    name: string,
    options?: DeprecationOptions
  ): () => void;
  function createDeprecation(name: string, options: DeprecationOptions = {}) {
    const errorAfter = getVersionFromOption(options.errorAfter);
    const warnAfter = getVersionFromOption(options.warnAfter);
    const since = getVersionFromOption(options.since) ?? warnAfter;
    const shouldError =
      options.error ||
      (errorAfter &&
        generatorVersion.compareTo(errorAfter) === Comparison.GreaterThan);
    const shouldWarn =
      !warnAfter || generatorVersion.compareTo(warnAfter) >= Comparison.Equal;

    return shouldError
      ? createErrorDeprecation(name, errorAfter, since, options.message)
      : shouldWarn
      ? createWarningDeprecation(name, errorAfter, since, options.message)
      : noopFunction;

    function getVersionFromOption(option: Semver | string | undefined) {
      if (option) {
        return typeof option === "string" ? new Semver(option) : option;
      }
      return undefined;
    }
  }

  function wrapFunction<F extends AnyFunction>(
    deprecation: () => void,
    func: F
  ): F {
    return function (this: unknown) {
      deprecation();
      // @ts-ignore
      return func.apply(this, arguments); // eslint-disable-line prefer-rest-params
    } as F;
  }

  export function deprecate<F extends AnyFunction>(
    func: F,
    options?: DeprecationOptions
  ): F {
    const deprecation = createDeprecation(
      options?.name ?? getFunctionName(func),
      options
    );
    return wrapFunction(deprecation, func);
  }
}
