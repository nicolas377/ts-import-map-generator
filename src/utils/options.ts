import { Debug } from "./debug";
import { keysOfObject } from "./generalHelpers";
import { ArgumentKind, ArgumentType, idToDataMap } from "cli/options/arguments";

const argumentKindToOptionNameMap: Readonly<
  Record<ArgumentKind, keyof IOptions>
> = {
  [ArgumentKind.Help]: "printHelpAndExit",
  [ArgumentKind.Version]: "printVersionAndExit",
  [ArgumentKind.Entrypoint]: "entrypointLocation",
  [ArgumentKind.GraphMaxDepth]: "graphMaxDepth",
  [ArgumentKind.IgnoreFiles]: "ignoreFiles",
};

interface IOptions {
  getOption(kind: ArgumentKind): ArgumentType;
  setOption(kind: ArgumentKind, value: ArgumentType): void;
  printHelpAndExit: boolean;
  printVersionAndExit: boolean;
  entrypointLocation: string;
  graphMaxDepth: number;
  ignoreFiles: string;
}

class OptionsClass {
  private allOptions = keysOfObject(argumentKindToOptionNameMap);

  private options: Record<ArgumentKind, ArgumentType>;

  public getOption(kind: ArgumentKind): ArgumentType {
    return this.options[kind];
  }

  public setOption(kind: ArgumentKind, value: ArgumentType): void {
    this.options[kind] = value;
  }

  constructor() {
    // kind should be one of all of the options, so this validates that at compile-time.
    this.options = this.allOptions.reduce((acc, kind: ArgumentKind) => {
      const { defaultValue } = idToDataMap.get(+kind) ?? {};

      Debug.assertIsDefined(
        defaultValue,
        `No default value for argument kind ${kind}`
      );

      acc[kind] = defaultValue;
      return acc;
    }, {} as Record<ArgumentKind, ArgumentType>);

    for (const kind of this.allOptions) {
      const optionData = idToDataMap.get(+kind);

      Debug.assertIsDefined(
        optionData,
        `No option data for argument kind ${kind}`
      );

      Object.defineProperty(this, argumentKindToOptionNameMap[kind], {
        configurable: false,
        enumerable: true,
        get: () => this.options[kind],
        set: (value: ArgumentType) => {
          // TODO: when TS supports merging function types, this assertion can be removed.
          (optionData.validator as (value: ArgumentType) => boolean)(value);
          this.options[kind] = value;
        },
      });
    }
  }
}

export const programOptions = new OptionsClass() as unknown as IOptions;
