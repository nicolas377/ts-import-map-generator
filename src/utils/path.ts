import {
  readdirSync as readDirectory,
  readFileSync as readFile,
  statSync,
} from "node:fs";
import {
  relative as getRelativePath,
  resolve as getResolvedPath,
  join as joinPaths,
  isAbsolute as pathIsAbsolute,
} from "node:path/posix";
import { cwd as getCurrentWorkingDir } from "node:process";
import { Debug } from "./debug";

export const enum PathType {
  DoesNotExist = 1 << 0,
  File = 1 << 1,
  Directory = 1 << 2,
}

// It doesn't matter what the relativeTo is, just that it's consistent.
const relativeTo = getCurrentWorkingDir();

interface IPathData {
  expandedPath: string;
  relativePath: string;
  type: PathType;
  exists: boolean;
}

interface IPath {
  type: PathType;
  exists: boolean;
  readFile(): string;
  readDir(): Path[];
}

function getPathData(path: string): IPathData {
  const fileStats = statSync(path);

  let isDirectory = false;
  const pathExists =
    fileStats.isFile() || (fileStats.isDirectory() && (isDirectory = true));
  const expandedPath = pathIsAbsolute(path)
    ? getResolvedPath(path)
    : getResolvedPath(relativeTo, path);
  const relativePath = getRelativePath(relativeTo, expandedPath);

  return {
    expandedPath,
    relativePath,
    type: getTypeOfFile(),
    exists: pathExists,
  };

  function getTypeOfFile(): PathType {
    if (!pathExists) {
      return PathType.DoesNotExist;
    }
    if (isDirectory) {
      return PathType.Directory;
    }
    return PathType.File;
  }
}

export class Path implements IPath {
  public readonly type: PathType;
  public readonly exists: boolean;
  public readonly relativePath: string;
  public readonly expandedPath: string;

  constructor(path: string) {
    // Empty path is not allowed
    Debug.assert(path !== "", "Path cannot be empty");

    const pathData = getPathData(path);

    this.expandedPath = pathData.expandedPath;
    this.type = pathData.type;
    this.exists = pathData.exists;
    this.relativePath = pathData.relativePath;
  }

  public readFile(): string {
    // checking bitflags to ensure that the path is a file
    Debug.assert(
      this.type & (PathType.DoesNotExist | PathType.Directory),
      "Path is not a file"
    );

    return readFile(this.expandedPath, "utf8");
  }

  public readDir(): Path[] {
    // checking bitflags to ensure that the path is a directory
    Debug.assert(
      this.type & (PathType.DoesNotExist | PathType.File),
      "Path is not a directory"
    );

    const paths: Path[] = [];
    // these strings are paths in the directory
    const directoryContents = readDirectory(this.expandedPath);

    for (const path of directoryContents) {
      const realPath = joinPaths(this.expandedPath, path);
      paths.push(new Path(realPath));
    }

    return paths;
  }
}
