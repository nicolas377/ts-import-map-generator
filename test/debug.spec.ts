import { getErrorFromThrowingFunction } from "./helpers";
import { Debug, LoggingHost, LogLevel } from "utils/debug";
import { Semver, setGeneratorVersion } from "version";

interface Log {
  level: LogLevel;
  message: string;
}

describe("Debug logging", () => {
  let logs: Log[] = [];
  const logHost: LoggingHost = {
    log: (level, message) => {
      logs.push({ level, message });
    },
  };

  beforeEach(() => {
    Debug.loggingHost = logHost;
  });

  afterEach(() => {
    Debug.currentLogLevel = LogLevel.Warning;
    logs = [];
  });

  test("only logs equal to or above the current log level", () => {
    Debug.currentLogLevel = LogLevel.Info;

    Debug.log.error("error");
    Debug.log.warn("warn");
    Debug.log.info("info");
    Debug.log.trace("trace");

    expect(logs).toEqual([
      { level: LogLevel.Error, message: "error" },
      { level: LogLevel.Warning, message: "warn" },
      { level: LogLevel.Info, message: "info" },
    ]);
  });

  test("log level changes don't affect previous logs", () => {
    Debug.currentLogLevel = LogLevel.Warning;

    Debug.log.error("error");
    Debug.log.warn("warn");
    Debug.log.info("info");
    Debug.log.trace("trace");

    Debug.currentLogLevel = LogLevel.Info;

    Debug.log.error("error");
    Debug.log.warn("warn");
    Debug.log.info("info");
    Debug.log.trace("trace");

    expect(logs).toEqual([
      { level: LogLevel.Error, message: "error" },
      { level: LogLevel.Warning, message: "warn" },
      { level: LogLevel.Error, message: "error" },
      { level: LogLevel.Warning, message: "warn" },
      { level: LogLevel.Info, message: "info" },
    ]);
  });

  test("calls log host for every log that should be logged", () => {
    const logMock = jest.fn();

    Debug.currentLogLevel = LogLevel.Info;
    Debug.loggingHost = { log: logMock };

    Debug.log.error("error");
    Debug.log.warn("warn");
    Debug.log.info("info");
    Debug.log.trace("trace");

    expect(logMock).toHaveBeenCalledTimes(3);
  });
});

describe("Debug assertions", () => {
  test.each(["message", "ensuring that it doesn't throw a constant message"])(
    "fail throws an error with the given message",
    (message) => {
      expect(() => Debug.fail(message)).toThrow(message);
    }
  );

  test("takes a function that originated the error and generates the stack trace off of it", () => {
    // There isn't a great way to test this, since stack traces differ all the time
    // We can at least make sure that the stack trace is different than if no function was provided

    const errorWithNothingProvided = getErrorFromThrowingFunction(() =>
      Debug.fail("message")
    );
    const errorWithFunctionProvided = getErrorFromThrowingFunction(() =>
      // eslint-disable-next-line @typescript-eslint/no-empty-function
      Debug.fail("message", () => {})
    );

    expect(errorWithNothingProvided.stack).not.toEqual(
      errorWithFunctionProvided.stack
    );
  });

  test("assert ensures that the condition is true", () => {
    expect(() => Debug.assert(true)).not.toThrow();
    expect(() => Debug.assert(false)).toThrow();
  });

  const assertEqualTests: [a: unknown, b: unknown][] = [
    [1, 1],
    [1, 2],
    [true, true],
    [true, false],
    [false, false],
    [false, true],
    ["a", "a"],
    ["a", "b"],
    [undefined, undefined],
    [undefined, null],
    [null, null],
    [null, undefined],
    [NaN, NaN],
    [NaN, 0],
    [0, NaN],
    [Infinity, Infinity],
    [Infinity, -Infinity],
    [-Infinity, -Infinity],
    [-Infinity, Infinity],
    [{}, {}],
    [{ a: 1 }, { b: 2 }],
  ];

  test.each(assertEqualTests)(
    "assertEqual ensures the two passed values are equal (using ===) %p === %p",
    (a, b) => {
      // assertEqual uses value checking, not deep equality
      /* eslint-disable jest/no-conditional-expect */
      if (Object.is(a, b)) {
        expect(() => Debug.assertEqual(a, b)).not.toThrow();
      } else {
        expect(() => Debug.assertEqual(a, b)).toThrow();
      }
      /* eslint-enable jest/no-conditional-expect */
    }
  );

  test("assertLessThan ensures the first value is less than the second", () => {
    expect(() => Debug.assertLessThan(1, 2)).not.toThrow();
    expect(() => Debug.assertLessThan(2, 1)).toThrow();
  });

  test("assertLessThanOrEqual ensures the first value is less than or equal to the second", () => {
    expect(() => Debug.assertLessThanOrEqual(1, 2)).not.toThrow();
    expect(() => Debug.assertLessThanOrEqual(2, 1)).toThrow();
    expect(() => Debug.assertLessThanOrEqual(1, 1)).not.toThrow();
  });

  test("assertGreaterThan ensures the first value is greater than the second", () => {
    expect(() => Debug.assertGreaterThan(2, 1)).not.toThrow();
    expect(() => Debug.assertGreaterThan(1, 2)).toThrow();
  });

  test("assertGreaterThanOrEqual ensures the first value is greater than or equal to the second", () => {
    expect(() => Debug.assertGreaterThanOrEqual(2, 1)).not.toThrow();
    expect(() => Debug.assertGreaterThanOrEqual(1, 2)).toThrow();
    expect(() => Debug.assertGreaterThanOrEqual(1, 1)).not.toThrow();
  });

  test("assertIsDefined ensures the value is not undefined", () => {
    expect(() => Debug.assertIsDefined(1)).not.toThrow();
    expect(() => Debug.assertIsDefined(undefined)).toThrow();
  });

  test("checkDefined only returns if the value is defined", () => {
    expect(Debug.checkDefined(1)).toBe(1);
    expect(() => Debug.checkDefined(1)).not.toThrow();
    expect(() => Debug.checkDefined(undefined)).toThrow();
  });

  test("assertEachIsDefined ensures each value of an array is defined", () => {
    expect(() => Debug.assertEachIsDefined([1, 2])).not.toThrow();
    expect(() => Debug.assertEachIsDefined([1, undefined])).toThrow();
  });

  // assertNever is for type-checking logic when CFA narrowing is involved
  test("assertNever throws", () => {
    expect(() => Debug.assertNever("" as never)).toThrow();
  });
});

describe("Debug deprecations", () => {
  let logs: Log[] = [];

  const beforeEachHook = () => {
    Debug.currentLogLevel = LogLevel.Warning;
    Debug.loggingHost = {
      log: (level, message) => {
        logs.push({ level, message });
      },
    };
    logs = [];
    setGeneratorVersion(new Semver("3.2.7"));
  };

  beforeEach(beforeEachHook);

  test("deprecation does nothing by default", () => {
    const deprecatedFunction = Debug.deprecate(() => "function result");

    expect(deprecatedFunction()).toBe("function result");
    expect(deprecatedFunction).not.toThrow();
    expect(logs).toBeEmpty();
  });

  test("warning deprecation still runs the function", () => {
    // ensuring that types don't get messed up by the deprecation function
    const warnDeprecatedFunction: () => string = Debug.deprecate(
      function someDeprecatedFunction() {
        return "function result";
      },
      {
        message: "function is deprecated",
        warnAfter: "2.0.0",
      }
    );

    expect(warnDeprecatedFunction()).toBe("function result");
  });

  test("warning deprecation logs a warning", () => {
    const warnDeprecatedFunction = Debug.deprecate(
      function someDeprecatedFunction() {
        return "function result";
      },
      {
        message: "function is deprecated",
        warnAfter: "2.0.0",
      }
    );

    warnDeprecatedFunction();

    expect(logs).toEqual([
      {
        level: LogLevel.Warning,
        message: expect.stringContaining("function is deprecated"),
      },
    ]);
    expect(logs[0]?.message).toMatchSnapshot();
  });

  const deprecationTests: [
    options: Debug.DeprecationOptions,
    warn: boolean,
    error: boolean
  ][] = [
    [{}, false, false],
    [{ warnAfter: "2.0.0" }, true, false],
    [{ message: "function is deprecated" }, false, false],
  ];

  test.each(deprecationTests)(
    "check errors %o",
    (deprecationOptions, _shouldWarn, shouldError) => {
      const warnDeprecatedFunction = Debug.deprecate(
        function deprecatedFunction() {
          return "function result";
        },
        deprecationOptions
      );

      /* eslint-disable jest/no-conditional-expect */
      if (shouldError) {
        expect(warnDeprecatedFunction).toThrowErrorMatchingSnapshot(
          "error snapshot"
        );
      } else {
        expect(warnDeprecatedFunction).not.toThrow();
      }
    }
  );

  test.each(deprecationTests)(
    "check warnings %o",
    (deprecationOptions, shouldWarn) => {
      const warnDeprecatedFunction = Debug.deprecate(
        function deprecatedFunction() {
          return "function result";
        },
        deprecationOptions
      );

      try {
        warnDeprecatedFunction();
      } catch (e) {
        // ignore
      }

      if (shouldWarn) {
        expect(logs).toEqual([
          {
            level: LogLevel.Warning,
            message: expect.stringContaining("function is deprecated"),
          },
        ]);
        expect(logs[0]?.message).toMatchSnapshot("warning snapshot");
      } else {
        expect(logs).toBeEmpty();
      }
    }
  );
  /* eslint-enable jest/no-conditional-expect */

  test("doesn't error before given version", () => {
    setGeneratorVersion(new Semver("1.0.0"));
    const warnDeprecatedFunction = Debug.deprecate(
      function deprecatedFunction() {
        return "function result";
      },
      {
        errorAfter: "2.0.0",
      }
    );

    // Should also not warn
    expect(logs).toBeEmpty();
    expect(warnDeprecatedFunction).not.toThrow();
  });

  test("errors after given version", () => {
    setGeneratorVersion(new Semver("3.0.0"));
    const warnDeprecatedFunction = Debug.deprecate(
      function deprecatedFunction() {
        return "function result";
      },
      {
        errorAfter: "2.0.0",
      }
    );

    expect(warnDeprecatedFunction).toThrowErrorMatchingSnapshot(
      "error snapshot"
    );
  });

  test("warn + error combo (post-warn)", () => {
    setGeneratorVersion(new Semver("2.1.0"));
    const warnDeprecatedFunction = Debug.deprecate(
      function deprecatedFunction() {
        return "function result";
      },
      {
        warnAfter: "2.0.0",
        errorAfter: "3.0.0",
      }
    );

    warnDeprecatedFunction();

    expect(logs).toEqual([
      {
        level: LogLevel.Warning,
        message: expect.toBeString(),
      },
    ]);
    expect(logs[0]?.message).toMatchSnapshot("warning snapshot");
  });

  test("warn + error combo (post-error)", () => {
    setGeneratorVersion(new Semver("3.1.0"));
    const warnDeprecatedFunction = Debug.deprecate(
      function deprecatedFunction() {
        return "function result";
      },
      {
        warnAfter: "2.0.0",
        errorAfter: "3.0.0",
      }
    );

    expect(warnDeprecatedFunction).toThrowErrorMatchingSnapshot(
      "error snapshot"
    );
  });

  test("warn + error combo (pre-warn)", () => {
    setGeneratorVersion(new Semver("1.0.0"));
    const warnDeprecatedFunction = Debug.deprecate(
      function deprecatedFunction() {
        return "function result";
      },
      {
        warnAfter: "2.0.0",
        errorAfter: "3.0.0",
      }
    );

    expect(warnDeprecatedFunction).not.toThrow();
    expect(logs).toBeEmpty();
  });
});
