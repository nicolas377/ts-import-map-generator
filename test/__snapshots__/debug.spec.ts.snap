// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debug deprecations errors after given version: error snapshot 1`] = `"DeprecationError: 'deprecatedFunction' is deprecated and can no longer be used."`;

exports[`Debug deprecations warn + error combo (post-error): error snapshot 1`] = `"DeprecationError: 'deprecatedFunction' has been deprecated since v2.0.0 and can no longer be used."`;

exports[`Debug deprecations warn + error combo (post-warn): warning snapshot 1`] = `"DeprecationWarning: 'deprecatedFunction' has been deprecated since v2.0.0 and will no longer be usable after v3.0.0"`;

exports[`Debug deprecations warning deprecation logs a warning 1`] = `"DeprecationWarning: 'someDeprecatedFunction' has been deprecated since v2.0.0. function is deprecated"`;
